virttest/ip_sniffing.py

```diff
@@ -0,0 +1,272 @@
+        logging.debug("Clean out all the address caches")
+
+
+class Sniffer(object):
+
+    """
+    Virtual base class of the ip sniffer abstraction.
+
+    The `_output_handler(self, line)` method must be
+    provided by the subclasses.
+    """
+
+    #: Sniffer's command name
+    command = ""
+    #: Sniffer's options
+    options = ""
+
+    def __init__(self, addr_cache, log_file, remote_opts=None):
+        """
+        Initializes the sniffer.
+
+        :param addr_cache: Address cache to be updated.
+        :param log_file: Log file name.
+        :param remote_opts: Options of remote host session, is a tuple built
+                            up by `(address, port, username, password,
+                            prompt)`. If provided, launches the sniffer on
+                            remote host, otherwise on local host.
+        """
+        self._cache = addr_cache
+        self._logfile = log_file
+        self._remote_opts = remote_opts
+        self._process = None
+
+    @classmethod
+    def _is_supported_remote(cls, session):
+        return session.cmd_status("which %s" % cls.command) == 0
+
+    @classmethod
+    def is_supported(cls, session=None):
+        """
+        Check if host supports the sniffer.
+
+        :param session: Remote host session. If provided, performs the
+                        check on remote host, otherwise on local host.
+        """
+        if session:
+            return cls._is_supported_remote(session)
+        try:
+            utils_path.find_command(cls.command)
+            return True
+        except utils_path.CmdNotFoundError:
+            return False
+
+    def _output_handler(self, line):
+        raise NotImplementedError
+
+    def _output_logger_handler(self, line):
+        try:
+            log_line(self._logfile, line)
+        except Exception as e:
+            logging.warn("Can't log ip sniffer output: '%s'", e)
+        self._output_handler(line)
+
+    def _start_remote(self):
+        address, port, username, password, prompt = self._remote_opts
+        cmd = "%s %s" % (self.command, self.options)
+        logging.debug("Run '%s' on host '%s'", cmd, address)
+        login_cmd = ("ssh -o UserKnownHostsFile=/dev/null "
+                     "-o StrictHostKeyChecking=no "
+                     "-o PreferredAuthentications=password -p %s %s@%s" %
+                     (port, username, address))
+
+        self._process = aexpect.ShellSession(
+            login_cmd,
+            output_func=self._output_logger_handler)
+        handle_prompts(self._process, username, password, prompt)
+        self._process.sendline(cmd)
+
+    def _start(self):
+        cmd = "%s %s" % (self.command, self.options)
+        if self._remote_opts:
+            self._start_remote()
+        else:
+            self._process = aexpect.run_tail(
+                command=cmd,
+                output_func=self._output_logger_handler)
+
+        if not self._process.is_alive():
```

Comment:
Checking immediately after exec is not really accurate. How about adding `0.5`s sleep? (tshark reported failure and finished but this function did not catch it as it was still running at this point)

Reply:
According to the implementation of `aexpect.run_tail` (L188), it does (equally) have a `1`s sleep before executing the current line, so this turns out that it needs more time to check whether the process is terminated. I'll make it wait `2`s to try some luck (as I said in https://github.com/avocado-framework/avocado-vt/pull/1231#issuecomment-347736207, it is also defective).

Reply:
I just realized that there was a status reporting log in the output stream if the process had terminated, so maybe we can do some hack about that in `_output_handler`?